server/game/ChessClock.js

```diff
@@ -0,0 +1,98 @@
+class ChessClock {
+    constructor(player, time, delayToStartClock) {
+        this.player = player;
+        this.mainTime = time * 60;
+        this.timeLeft = time * 60;
+        this.mode = 'inactive';
+        this.timerStart = 0; //when the clock is running, this contains the instant on which it started
+        this.paused = false; //has the clock been paused via chat command?
+        this.running = false; //is the clock ticking for the player?
+        this.stateId = 0;
+        this.name = 'ChessClock';
+        this.activated = false; //indicates if the game has started/setup is finished 
+        this.delayToStartClock = delayToStartClock;
+
+        this.player.game.on('onSetupFinished', () => this.activateChessClock());
+    }
+
+    activateChessClock() {
+        this.activated = true;
+    }
+
+    togglePause() {
+        if(this.activated) {
+            this.paused = !this.paused;
+            if(this.paused) {
+                this.stop();
+            }
+        }
+    }
+
+    modify(secs) {
+        this.timeLeft += secs;
+    }
+
+    updateStateId() {
+        this.stateId++;
+    }
+
+    start() {
+        if(this.activated) {
+            if(!this.paused && !this.running) {

```

Comment:
Small stylistic choice, but perhaps early returns would be easier to read than these nested `if` statements?

Reply:
good point, fixed